File: backends/spring-java/src/main/java/dev/stackverse/backend/account/AdminUserController.java
package dev.stackverse.backend.account;

import static dev.stackverse.backend.common.RequestValidation.requireMaxLength;
import static dev.stackverse.backend.common.RequestValidation.requireValidPaging;

import com.fasterxml.jackson.annotation.JsonInclude;
import dev.stackverse.backend.common.BadRequestProblem;
import dev.stackverse.backend.common.NotFoundProblem;
import dev.stackverse.backend.common.PageResponse;
import java.time.Instant;
import java.util.Locale;
import org.springframework.data.domain.Pageable;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

record UserStatusRequest(UserAccountStatus status, String reason) {
}

@JsonInclude(JsonInclude.Include.NON_NULL)
record UserAccountResponse(
    String username,
    Instant firstSeen,
    Instant lastSeen,
    UserAccountStatus status,
    String blockedReason,
    long bookmarkCount
) {
    static UserAccountResponse of(UserAccount account, long bookmarkCount) {
        return new UserAccountResponse(
            account.getUsername(),
            account.getFirstSeen(),
            account.getLastSeen(),
            account.getStatus(),
            account.getBlockedReason(),
            bookmarkCount
        );
    }
}

@RestController
@RequestMapping("/api/v1/admin/users")
@PreAuthorize("hasRole('admin')")
public class AdminUserController {
    private final UserAccountRepository repository;
    private final UserAccountService service;

    public AdminUserController(UserAccountRepository repository, UserAccountService service) {
        this.repository = repository;
        this.service = service;
    }

    @GetMapping
    public PageResponse<UserAccountResponse> list(
        @RequestParam(name = "q", required = false) String q,
        @RequestParam(name = "status", required = false) UserAccountStatus status,
        @RequestParam(name = "page", defaultValue = "0") int page,
        @RequestParam(name = "size", defaultValue = "20") int size
    ) {
        requireValidPaging(page, size);
        requireMaxLength(q, 100, "q");
        String qLike = q == null || q.isBlank() ? null : "%" + escapeLike(q.toLowerCase(Locale.ROOT)) + "%";
        return PageResponse.of(repository.search(qLike, status, Pageable.ofSize(size).withPage(page)), row ->
            UserAccountResponse.of(row.getAccount(), row.getBookmarkCount())
        );
    }

    @GetMapping("/{username}")
    public UserAccountResponse get(@PathVariable("username") String username) {
        AccountWithBookmarkCount row = repository.findWithBookmarkCount(username);
        if (row == null) {
            throw new NotFoundProblem();
        }
        return UserAccountResponse.of(row.getAccount(), row.getBookmarkCount());
    }

    @PutMapping("/{username}/status")
    public UserAccountResponse setStatus(
        @PathVariable("username") String username,
        @RequestBody UserStatusRequest request,
        Authentication authentication
    ) {
        if (request.status() == null) {
            throw new BadRequestProblem("status is required");
        }
        UserAccount account = service.setStatus(authentication.getName(), username, request.status(), trimToNull(request.reason()));
        AccountWithBookmarkCount row = repository.findWithBookmarkCount(username);
        if (row == null) {
            throw new NotFoundProblem();
        }
        return UserAccountResponse.of(account, row.getBookmarkCount());
    }

    private static String trimToNull(String value) {
        return value == null ? null : value.trim();
    }

    private static String escapeLike(String value) {
        return value.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_");
    }
}

File: backends/spring-kotlin/src/main/kotlin/dev/stackverse/backend/account/AdminUserController.kt
package dev.stackverse.backend.account

import com.fasterxml.jackson.annotation.JsonInclude
import dev.stackverse.backend.common.BadRequestProblem
import dev.stackverse.backend.common.NotFoundProblem
import dev.stackverse.backend.common.PageResponse
import dev.stackverse.backend.common.requireValidPaging
import org.springframework.data.domain.Pageable
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.security.core.Authentication
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.time.Instant

data class UserStatusRequest(val status: UserAccountStatus? = null, val reason: String? = null)

@JsonInclude(JsonInclude.Include.NON_NULL)
data class UserAccountResponse(
    val username: String,
    val firstSeen: Instant,
    val lastSeen: Instant,
    val status: UserAccountStatus,
    val blockedReason: String?,
    val bookmarkCount: Long,
) {
    companion object {
        fun of(account: UserAccount, bookmarkCount: Long) = UserAccountResponse(
            username = account.username,
            firstSeen = account.firstSeen,
            lastSeen = account.lastSeen,
            status = account.status,
            blockedReason = account.blockedReason,
            bookmarkCount = bookmarkCount,
        )
    }
}

@RestController
@RequestMapping("/api/v1/admin/users")
@PreAuthorize("hasRole('admin')")
class AdminUserController(
    private val repository: UserAccountRepository,
    private val service: UserAccountService,
) {

    @GetMapping
    fun list(
        @RequestParam q: String?,
        @RequestParam status: UserAccountStatus?,
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "20") size: Int,
    ): PageResponse<UserAccountResponse> {
        requireValidPaging(page, size)
        val qLike = q?.takeIf { it.isNotBlank() }?.let { "%${escapeLike(it.lowercase())}%" }
        val result = repository.search(qLike, status, Pageable.ofSize(size).withPage(page))
        return PageResponse.of(result) { UserAccountResponse.of(it.account, it.bookmarkCount) }
    }

    @GetMapping("/{username}")
    fun get(@PathVariable username: String): UserAccountResponse {
        val row = repository.findWithBookmarkCount(username) ?: throw NotFoundProblem()
        return UserAccountResponse.of(row.account, row.bookmarkCount)
    }

    @PutMapping("/{username}/status")
    fun setStatus(
        @PathVariable username: String,
        @RequestBody request: UserStatusRequest,
        authentication: Authentication,
    ): UserAccountResponse {
        val status = request.status ?: throw BadRequestProblem("status is required")
        val account = service.setStatus(authentication.name, username, status, request.reason?.trim())
        val row = repository.findWithBookmarkCount(username) ?: throw NotFoundProblem()
        return UserAccountResponse.of(account, row.bookmarkCount)
    }

    private fun escapeLike(value: String): String =
        value.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")
}
